pkg/resultstore/logline.go

```diff
@@ -54,9 +55,18 @@ func (s *store) openAllLogReadersForCatchingUp() ([]LogLineReadCloser, error) {
 	return readers, nil
 }
 
-func (s *store) pubAllLogsToChanToCatchUp(readers []LogLineReadCloser, pubSub *chans.PubSub[LogLine]) {
+func (s *store) pubAllLogsToChanToCatchUp(readers []LogLineReadCloser, pubSub *chans.PubSub[LogLine], ch <-chan LogLine) {
+	wg := sync.WaitGroup{}
+	wg.Add(len(readers))
 	for _, r := range readers {
-		go s.pubLogsToChanToCatchUp(r, pubSub)
+		go func(r LogLineReadCloser) {
+			s.pubLogsToChanToCatchUp(r, pubSub)
+			wg.Done()
+		}(r)
+	}
+	wg.Wait()
+	if s.frozen {
+		s.logPubSub.Unsub(ch)
 	}
 }
```

Comment:
This freezing logic is messy. If it works, that's great. But a rewrite is in order.

Let's postpone such a rewrite for a different PR though.

My idea is that we should be able to reuse the readers for the entire reading process, and let all messages get piped through the file. In other words:

- new log line comes in from pod
- write log line to file
- (wake up any sleeping readers)
- reader continues to read and publishes all log lines via pubsub, reads until reaches EOF, and then goes to sleep
- on freeze, wake up all readers and tell them to stop on EOF instead of go to sleep.

For wake up/sleep we could use channels. Ex:

```go
func stuff() {
	sleepCh := make(chan struct{})

	// reader:
	go func() {
		for {
			// 1. read until EOF, and send all logs through pubsub
			// 2. check if should die or sleep
			<-sleepCh
		}
	}()

	// writer:
	go func() {
		for {
			// 1. wait for log line from pod
			// 2. write log line to file
			wakeUp(sleepCh)
		}
	}()
}

func wakeUp(ch chan struct{}) {
	select {
	case ch<-struct{}{}:
	default:
	}
}
```

Reply:
Created #82. Resolving.